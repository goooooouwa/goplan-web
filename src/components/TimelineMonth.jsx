import { Box, Grid, IconButton, Stack, Typography } from '@mui/material';
import { isInMonthRange } from 'lib/rangeCheck';
import moment from 'moment';
import React from 'react';
import { Link as RouterLink, useParams } from 'react-router-dom';
import TodoMonthSlider from './TodoMonthSlider';

const marks = [
  {
    value: 1,
    label: 'Week 1',
  },
  {
    value: 2,
    label: 'Week 2',
  },
  {
    value: 3,
    label: 'Week 3',
  },
  {
    value: 4,
    label: 'Week 4',
  },
  {
    value: 5,
    label: 'Week 5',
  },
];

export default function TimelineMonth(props) {
  const params = useParams();
  const weekUrlPrefix = params.projectId !== undefined ? `/projects/${params.projectId}/week?week=` : '/timeline/week?week=';

  return (
    <>
      <Grid container rowSpacing={1} >
        <Grid item xs={12} md={4}>
          <Typography
            variant="body1"
            gutterBottom
            sx={{
              fontWeight: 'bold',
              textAlign: 'left',
              mt: 1,
              ml: 2
            }}
          >
            Todos
          </Typography>
        </Grid>
        <Grid item xs={12} md={8}>
          <Box m={1}>
            <Stack
              direction="row"
              justifyContent="space-between"
              alignItems="center"
            >
              {marks.map((mark, index) => (
                <Stack key={index}>
                  <Typography
                    variant="body2"
                    sx={{
                      color: (isInMonthRange(moment(), props.selectedMonth) && (Math.ceil(moment().date() / 7) === index)) ? 'error.main' : 'text.primary'
                    }}
                  >
                    {mark.label}
                  </Typography>
                  <IconButton
                    key={index}
                    component={RouterLink}
                    size="small"
                    to={`${weekUrlPrefix}${props.selectedMonth.clone().add(mark.value, "weeks").format("YYYY[W]WW")}`}
                    sx={{
                      fontWeight: 'bold',
                      fontSize: 16,
                      color: (isInMonthRange(moment(), props.selectedMonth) && (Math.ceil(moment().date() / 7) === index)) ? 'error.main' : 'text.primary'
                    }}
                  >
                    {`${props.selectedMonth.clone().startOf("month").add(mark.value - 1, "weeks").format("D")} - ${moment.min(props.selectedMonth.clone().startOf("month").add(mark.value - 1, "weeks").add(6, "days"), props.selectedMonth.clone().endOf("month")).format("D")}`}
                  </IconButton>
                </Stack>
              ))}
            </Stack>
          </Box>
        </Grid>
        {props.todos
          .sort((t1, t2) => moment(t1.createdAt).isBefore(t2.createdAt) ? -1 : 1)
          .map((todo, index) => (
            <TodoMonthSlider key={index} todo={todo} marks={marks} selectedMonth={props.selectedMonth} />
          ))}
      </Grid>
    </>
  );
}
