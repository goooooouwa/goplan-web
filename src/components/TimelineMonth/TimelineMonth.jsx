import { Box, Grid, IconButton, Stack, Typography } from '@mui/material';
import { isInMonthRange, nthWeekOfMonth, todosInMonthRange } from 'utils/rangeCheck';
import moment from 'moment';
import React from 'react';
import { Link as RouterLink, useParams } from 'react-router-dom';
import PropTypes from 'prop-types';
import momentPropTypes from 'react-moment-proptypes';
import SHARED_PROP_TYPES from 'utils/sharedPropTypes';
import TodoMonthSlider from '../TodoMonthSlider/TodoMonthSlider';
import { useTranslation } from 'react-i18next';

const marks = [
  {
    value: 1,
    label: 'Week 1',
  },
  {
    value: 2,
    label: 'Week 2',
  },
  {
    value: 3,
    label: 'Week 3',
  },
  {
    value: 4,
    label: 'Week 4',
  },
  {
    value: 5,
    label: 'Week 5',
  },
];

export default function TimelineMonth(props) {
  const { t, i18n } = useTranslation();
  const params = useParams();
  const weekUrlPrefix = params.projectId !== undefined ? `/projects/${params.projectId}/week?week=` : '/timeline/week?week=';

  return (
    <>
      <Grid container rowSpacing={1} >
        <Grid item xs={12} md={4}>
          <Typography
            variant="body1"
            gutterBottom
            sx={{
              fontWeight: 'bold',
              textAlign: 'left',
              mt: 1,
              ml: 2
            }}
          >
            {t('Tasks')}
          </Typography>
        </Grid>
        <Grid item xs={12} md={8}>
          <Box m={1}>
            <Stack
              direction="row"
              justifyContent="space-between"
              alignItems="center"
            >
              {marks.map((mark, index) => (
                <Stack key={index}>
                  <Typography
                    variant="body2"
                    sx={{
                      color: (isInMonthRange(moment(), props.selectedMonth) && (nthWeekOfMonth(moment(), props.selectedMonth) === index + 1)) ? 'error.main' : 'text.primary'
                    }}
                  >
                    {mark.label}
                  </Typography>
                  <IconButton
                    key={index}
                    component={RouterLink}
                    size="small"
                    to={`${weekUrlPrefix}${props.selectedMonth.clone().add(mark.value - 1, "weeks").format("YYYY-MM-DD")}`}
                    sx={{
                      fontWeight: 'bold',
                      fontSize: 16,
                      color: (isInMonthRange(moment(), props.selectedMonth) && (nthWeekOfMonth(moment(), props.selectedMonth) === index + 1)) ? 'error.main' : 'text.primary'
                    }}
                  >
                    {`${props.selectedMonth.clone().add(mark.value - 1, "weeks").startOf("week").format("D")} - ${props.selectedMonth.clone().add(mark.value - 1, "weeks").startOf("week").add(6, "days").format("D")}`}
                  </IconButton>
                </Stack>
              ))}
            </Stack>
          </Box>
        </Grid>
        {props.todos.map((todo, index) => (
          <TodoMonthSlider key={index} todo={todo} selectedMonth={props.selectedMonth} handleTodoChange={props.handleTodoChange} handleWeekChange={props.handleWeekChange} />
        ))}
        {props.todos.length === 0 &&
          <Grid item xs={12}>
            <Stack alignItems="center" justifyContent="center" sx={{ height: 200 }}>
              <Typography alignItems="center">
                {t('No tasks in selected time range')}
              </Typography>
            </Stack>
          </Grid>
        }
      </Grid>
    </>
  );
}

TimelineMonth.propTypes = {
  selectedMonth: momentPropTypes.momentObj.isRequired,
  todos: PropTypes.arrayOf(SHARED_PROP_TYPES.todo).isRequired,
  handleTodoChange: PropTypes.func.isRequired,
  handleWeekChange: PropTypes.func.isRequired
};