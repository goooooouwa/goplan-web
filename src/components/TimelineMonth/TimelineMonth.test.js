import React from 'react';
import renderer from 'react-test-renderer';
import moment from 'moment';
import { BrowserRouter } from 'react-router-dom';
import TimelineMonth from './TimelineMonth';

const todos = [{
  id: 1,
  projectId: 1,
  status: false,
  repeat: true,
  name: "Todo 1",
  startDate: '2022-01-02',
  endDate: '2022-02-02',
  instanceTimeSpan: 1,
}];

it('should show a timeline Month view', () => {
  const component = renderer.create(
    <BrowserRouter>
      <TimelineMonth todos={todos} selectedMonth={moment('2022-02-01')} handleTodoChange={() => { }} handleWeekChange={() => { }} />
    </BrowserRouter>
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});