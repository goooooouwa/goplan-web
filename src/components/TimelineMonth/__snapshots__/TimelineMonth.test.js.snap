// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should show a timeline Month view 1`] = `
<div
  className="MuiGrid-root MuiGrid-container css-1sjizby-MuiGrid-root"
>
  <div
    className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-4 css-1lj5egr-MuiGrid-root"
  >
    <p
      className="MuiTypography-root MuiTypography-body1 MuiTypography-gutterBottom css-1xzmp4b-MuiTypography-root"
    >
      Todos
    </p>
  </div>
  <div
    className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-8 css-9drkqd-MuiGrid-root"
  >
    <div
      className="MuiBox-root css-1ynyhby"
    >
      <div
        className="css-3y1137-MuiStack-root"
      >
        <div
          className="css-nen11g-MuiStack-root"
        >
          <p
            className="MuiTypography-root MuiTypography-body2 css-1yr8029-MuiTypography-root"
          >
            Week 1
          </p>
          <a
            className="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall css-1lhdua6-MuiButtonBase-root-MuiIconButton-root"
            href="/timeline/week?week=2022W05"
            onBlur={[Function]}
            onClick={[Function]}
            onContextMenu={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex={0}
          >
            30 - 5
          </a>
        </div>
        <div
          className="css-nen11g-MuiStack-root"
        >
          <p
            className="MuiTypography-root MuiTypography-body2 css-1yr8029-MuiTypography-root"
          >
            Week 2
          </p>
          <a
            className="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall css-1lhdua6-MuiButtonBase-root-MuiIconButton-root"
            href="/timeline/week?week=2022W06"
            onBlur={[Function]}
            onClick={[Function]}
            onContextMenu={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex={0}
          >
            6 - 12
          </a>
        </div>
        <div
          className="css-nen11g-MuiStack-root"
        >
          <p
            className="MuiTypography-root MuiTypography-body2 css-1yr8029-MuiTypography-root"
          >
            Week 3
          </p>
          <a
            className="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall css-1lhdua6-MuiButtonBase-root-MuiIconButton-root"
            href="/timeline/week?week=2022W07"
            onBlur={[Function]}
            onClick={[Function]}
            onContextMenu={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex={0}
          >
            13 - 19
          </a>
        </div>
        <div
          className="css-nen11g-MuiStack-root"
        >
          <p
            className="MuiTypography-root MuiTypography-body2 css-1yr8029-MuiTypography-root"
          >
            Week 4
          </p>
          <a
            className="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall css-1lhdua6-MuiButtonBase-root-MuiIconButton-root"
            href="/timeline/week?week=2022W08"
            onBlur={[Function]}
            onClick={[Function]}
            onContextMenu={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex={0}
          >
            20 - 26
          </a>
        </div>
        <div
          className="css-nen11g-MuiStack-root"
        >
          <p
            className="MuiTypography-root MuiTypography-body2 css-1yr8029-MuiTypography-root"
          >
            Week 5
          </p>
          <a
            className="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall css-1lhdua6-MuiButtonBase-root-MuiIconButton-root"
            href="/timeline/week?week=2022W09"
            onBlur={[Function]}
            onClick={[Function]}
            onContextMenu={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex={0}
          >
            27 - 5
          </a>
        </div>
      </div>
    </div>
  </div>
  <div
    className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-4 css-1lj5egr-MuiGrid-root"
  >
    <li
      className="MuiListItem-root MuiListItem-gutters css-1nxmd3h-MuiListItem-root"
      disabled={false}
    >
      <span
        className="MuiCheckbox-root MuiCheckbox-colorPrimary MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorPrimary PrivateSwitchBase-root css-12wnr2w-MuiButtonBase-root-MuiCheckbox-root"
        onBlur={[Function]}
        onContextMenu={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={null}
      >
        <input
          aria-label="controlled"
          checked={false}
          className="PrivateSwitchBase-input css-1m9pwf3"
          data-indeterminate={false}
          onChange={[Function]}
          type="checkbox"
        />
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="CheckBoxOutlineBlankIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
          />
        </svg>
      </span>
      <a
        className="MuiListItemButton-root MuiListItemButton-gutters MuiButtonBase-root css-16ac5r2-MuiButtonBase-root-MuiListItemButton-root"
        href="/projects/1/todos/1"
        onBlur={[Function]}
        onClick={[Function]}
        onContextMenu={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
      >
        <div
          className="MuiListItemText-root css-tlelie-MuiListItemText-root"
        >
          <span
            className="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
          >
            Todo 1
          </span>
        </div>
      </a>
    </li>
  </div>
  <div
    className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-8 css-1x9uoy6-MuiGrid-root"
  >
    <span
      className="MuiSlider-colorPrimary MuiSlider-sizeMedium MuiSlider-root css-187mznn-MuiSlider-root"
      onMouseDown={[Function]}
    >
      <span
        className="MuiSlider-rail css-14pt78w-MuiSlider-rail"
      />
      <span
        className="MuiSlider-track css-1gv0vcd-MuiSlider-track"
        style={
          Object {
            "left": "0%",
            "width": "0%",
          }
        }
      />
      <span
        className="MuiSlider-mark MuiSlider-markActive css-6cwnna-MuiSlider-mark"
        data-index={0}
        style={
          Object {
            "left": "0%",
          }
        }
      />
      <span
        className="MuiSlider-mark css-1tfve6q-MuiSlider-mark"
        data-index={1}
        style={
          Object {
            "left": "25%",
          }
        }
      />
      <span
        className="MuiSlider-mark css-1tfve6q-MuiSlider-mark"
        data-index={2}
        style={
          Object {
            "left": "50%",
          }
        }
      />
      <span
        className="MuiSlider-mark css-1tfve6q-MuiSlider-mark"
        data-index={3}
        style={
          Object {
            "left": "75%",
          }
        }
      />
      <span
        className="MuiSlider-mark css-1tfve6q-MuiSlider-mark"
        data-index={4}
        style={
          Object {
            "left": "100%",
          }
        }
      />
      <span
        className="MuiSlider-thumbColorPrimary MuiSlider-thumbSizeMedium MuiSlider-thumb css-eg0mwd-MuiSlider-thumb"
        data-index={0}
        onMouseLeave={[Function]}
        onMouseOver={[Function]}
        style={
          Object {
            "left": "0%",
            "pointerEvents": undefined,
          }
        }
      >
        <input
          aria-orientation="horizontal"
          aria-valuemax={5}
          aria-valuemin={1}
          aria-valuenow={1}
          aria-valuetext={1}
          data-index={0}
          disabled={false}
          max={5}
          min={1}
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          step={1}
          style={
            Object {
              "border": 0,
              "clip": "rect(0 0 0 0)",
              "direction": "ltr",
              "height": "100%",
              "margin": -1,
              "overflow": "hidden",
              "padding": 0,
              "position": "absolute",
              "whiteSpace": "nowrap",
              "width": "100%",
            }
          }
          type="range"
          value={1}
        />
        <span
          aria-hidden={true}
          className="MuiSlider-valueLabel css-1kz0hui-MuiSlider-valueLabel"
        >
          <span
            className="MuiSlider-valueLabelCircle"
          >
            <span
              className="MuiSlider-valueLabelLabel"
            >
              Week 1
            </span>
          </span>
        </span>
      </span>
      <span
        className="MuiSlider-thumbColorPrimary MuiSlider-thumbSizeMedium MuiSlider-thumb css-eg0mwd-MuiSlider-thumb"
        data-index={1}
        onMouseLeave={[Function]}
        onMouseOver={[Function]}
        style={
          Object {
            "left": "0%",
            "pointerEvents": undefined,
          }
        }
      >
        <input
          aria-orientation="horizontal"
          aria-valuemax={5}
          aria-valuemin={1}
          aria-valuenow={1}
          aria-valuetext={1}
          data-index={1}
          disabled={false}
          max={5}
          min={1}
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          step={1}
          style={
            Object {
              "border": 0,
              "clip": "rect(0 0 0 0)",
              "direction": "ltr",
              "height": "100%",
              "margin": -1,
              "overflow": "hidden",
              "padding": 0,
              "position": "absolute",
              "whiteSpace": "nowrap",
              "width": "100%",
            }
          }
          type="range"
          value={1}
        />
        <span
          aria-hidden={true}
          className="MuiSlider-valueLabel css-1kz0hui-MuiSlider-valueLabel"
        >
          <span
            className="MuiSlider-valueLabelCircle"
          >
            <span
              className="MuiSlider-valueLabelLabel"
            >
              Week 1
            </span>
          </span>
        </span>
      </span>
    </span>
  </div>
</div>
`;
